include/onnxruntime/core/framework/sparse_tensor.h

```diff
@@ -1,72 +1,427 @@
 
   // Returns the number of entries in the values tensor (aka "NNZ" or "number of nonzero values")
+  // For block sparse formats this may include some zeros in the blocks are considered non-zero.
   size_t NumValues() const { return static_cast<size_t>(values_.Shape().Size()); }
 
-  const Tensor& Values() const {
+  /// <summary>
+  /// Read only accessor to values
+  /// </summary>
+  /// <returns></returns>
+  const Tensor& Values() const noexcept {
     return values_;
   }
 
-  const Tensor& Indices() const {
-    return indices_;
+  SparseTensor(SparseTensor&& o) noexcept;
+  SparseTensor& operator=(SparseTensor&& o) noexcept;
+
+  /// <summary>
+  /// Returns SparseFormat that the instance currently holds
+  /// if the value returned in kUndefined, the instance is not populated
+  /// </summary>
+  /// <returns></returns>
+  SparseFormat Format() const noexcept {
+    return format_;
   }
 
-  const TensorShape& Shape() const {
-    return shape_;
+  /// <summary>
+  /// Returns a would be dense_shape
+  /// This does not describe shapes of values of indices or values.
+  /// </summary>
+  /// <returns></returns>
+  const TensorShape& Shape() const noexcept {
+    return dense_shape_;
   }
 
-  Tensor& MutableValues() {
-    return values_;
+  /// <summary>
+  /// Calculates and returns how much this fully initialized SparseTensor data (would)
+  /// occupy in a contiguous allocation block, or, in fact, occupies if it owns the buffer.
+  /// </summary>
+  /// <returns></returns>
+  int64_t RequiredAllocationSize() const noexcept;
+
+  /// <summary>
+  /// Returns Tensor element type enum.
+  /// Useful for type dispatching
+  /// </summary>
+  /// <returns></returns>
+  int32_t GetElementType() const {
+    return ml_data_type_->GetDataType();
   }
 
-  Tensor& MutableIndices() {
-    return indices_;
+  /// <summary>
+  /// Return Element MLDataType
+  /// </summary>
+  /// <returns></returns>
+  MLDataType DataType() const noexcept {
+    return ml_data_type_;
   }
 
-  //TensorShape& MutableShape() {
-  //  return shape_;
-  //}
+  /// <summary>
+  /// Test for string type
+  /// </summary>
+  /// <returns>true if tensor values are strings</returns>
+  bool IsDataTypeString() const {
+    return utils::IsPrimitiveDataType<std::string>(ml_data_type_);
+  }
+
+  /// <summary>
+  /// Checks if the Tensor contains data type T
+  /// </summary>
+  /// <typeparam name="T"></typeparam>
+  /// <returns>true if tensor contains data of type T</returns>
+  template <class T>
+  bool IsDataType() const {
+    return utils::IsPrimitiveDataType<T>(ml_data_type_);
+  }
+
+  const OrtMemoryInfo& Location() const noexcept { return location_; }
+
+  bool OwnsBuffer() const noexcept {
+    return allocator_ != nullptr;
+  }
+
+  /// <summary>
+  /// Read only access to Coo index
+  /// </summary>
+  class CooView {
+   public:
+    explicit CooView(const Tensor& index) noexcept
+        : index_(index) {}
```

Comment:
for the various view/mutator classes, should the ctor taking internal tensors be public? is it only meaningful to get a view/mutator instance from a SparseTensor or is there another case where one might call this?

Reply:
It is Okay for it to be public. If someone wants to get access to the actual data, it would not make sense to instantiate it with just any Tensor, one must use As*() methods.